### PR TITLE
Whereabouts IPAM CNI Sticky IP Addresses Enhancement

### PR DESCRIPTION
**Pull request by dougbtv**
This proposes a general design for how Whereabouts can handle sticky (persistent) IP addresses.

### CODE REVIEW

**Comment by itamarh** [comment]
There is another aspect around IPAM for providing same IP which came up when discussing KubeVirt and live migration for Pod networks. Could be very different/not in scope, but might be wroth considering when looking at this.

File: enhancements/network/whereabouts-sticky-ips.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+This feature implements functionality for Whereabouts where workloads are assigned an IP address, and when those workloads are terminated and restarted, are assigned the same IP address.
```
Comment (russellb) [inline enhancements/network/whereabouts-sticky-ips.md:29]: Is this an enhancement to something we already ship and support, or proposing Whereabouts in general?

If it's something we already ship, where could I read more about its inclusion in OpenShift and what use cases it solves today?

If it's introducing Whereabouts as a new thing, can you update this to do two things: first introduce the problem without mention of the solution (whereabouts).  Then can you summarize what wherabouts is, where it comes from, and how people can learn more?

Comment (danwinship) [reply enhancements/network/whereabouts-sticky-ips.md:29]: whereabouts is something we already ship, but that was pre-enhancements, and I have no idea what it does (although I'm sure Doug explained it to me at some point) so yeah, it would be good to have this enhancement as both an explanation of whereabouts _and_ an explanation of the new 4.7 feature on top of that.

Comment (russellb) [reply enhancements/network/whereabouts-sticky-ips.md:29]: +1 to explaining both, thanks :)

File: enhancements/network/whereabouts-sticky-ips.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+## Motivation
+
+This functionality has been requested to provide reliable IP addresses for workloads that advertise their IP address in a method other than Kubernetes services, and where external components require that those same IP addresses be reachable regardless if the workload has been restarted.
```
Comment (dhellmann) [inline enhancements/network/whereabouts-sticky-ips.md:33]: What does "advertise their IP address in a method other than Kubernetes services" mean?

File: enhancements/network/whereabouts-sticky-ips.md [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+(No relevant non-goals)
+
+## Proposal
+
+### Implementation Details/Notes/Constraints 
+
+Some upstream discussion regarding the design has already occured in [this upstream issue](https://github.com/dougbtv/whereabouts/issues/51).
+
+Generally the proposed design is to create a new CRD that is specific to an individual IP address. This can be used with, or as an extension to data that exists for storing IP addresses individually in order to determine if they are used in overlapping ranges.
```
Comment (dhellmann) [inline enhancements/network/whereabouts-sticky-ips.md:51]: CRD or CR?

It would be interesting to see how closely that matches https://github.com/metal3-io/metal3-docs/blob/master/design/ip-address-manager/ip-address-management-for-networkdata.md
